File: scripts/lint.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
HarnessWiki 기계 검증 스크립트 (Lint)

검증 항목:
1. Frontmatter 유효성 (YAML 포맷)
2. 스키마 구조 검증 (경로, 파일명)
3. 파일명 규칙 (kebab-case)
4. 링크 검증 (존재 여부, 경로 깊이)
"""

import os
import re
import sys
from pathlib import Path
from typing import List, Dict, Tuple
import yaml
import io

# Windows 인코딩 처리
if sys.platform == "win32":
    sys.stdout = io.TextIOWrapper(sys.stdout.buffer, encoding='utf-8')

class WikiLinter:
    def __init__(self, wiki_root: str = "."):
        self.wiki_root = Path(wiki_root)
        self.errors = []
        self.warnings = []
        self.info = []

    def log_error(self, file_path: str, message: str):
        self.errors.append(f"[ERROR] {file_path}: {message}")

    def log_warning(self, file_path: str, message: str):
        self.warnings.append(f"[WARN]  {file_path}: {message}")

    def log_info(self, file_path: str, message: str):
        self.info.append(f"[INFO]  {file_path}: {message}")

    def is_kebab_case(self, text: str) -> bool:
        """슬러그가 kebab-case 형식인지 확인"""
        # 허용: 소문자, 숫자, 하이픈, 한글, 기타 유니코드 문자
        # 한글은 kebab-case의 예외 허용
        if not text or text.startswith("-") or text.endswith("-"):
            return False
        # 한글 포함 확인
        if re.search(r'[一-鿿가-힯]', text):
            return True  # 한글은 허용
        # 영문+숫자+하이픈만 허용
        return bool(re.match(r"^[a-z0-9-]+$", text))

    def validate_raw_structure(self):
        """raw/ 폴더 구조 검증"""
        raw_dir = self.wiki_root / "raw"
        if not raw_dir.exists():
            self.log_warning("raw/", "디렉토리가 없음")
            return

        for folder in raw_dir.iterdir():
            if not folder.is_dir():
                continue

            folder_name = folder.name
            # YYYY-MM-DD_슬러그 형식 검증
            match = re.match(r"^(\d{4})-(\d{2})-(\d{2})_(.+)$", folder_name)
            if not match:
                self.log_error(f"raw/{folder_name}", "폴더명 형식 오류: YYYY-MM-DD_슬러그 형식 필요")
                continue

            slug = match.group(4)
            if not self.is_kebab_case(slug):
                self.log_error(f"raw/{folder_name}", f"슬러그 '{slug}'가 kebab-case 아님 (허용: 소문자, 숫자, 하이픈)")

            # raw.md, meta.md 존재 확인
            raw_file = folder / "raw.md"
            meta_file = folder / "meta.md"

            if not raw_file.exists():
                self.log_error(f"raw/{folder_name}", "raw.md 파일 없음")
            else:
                self.validate_frontmatter(str(raw_file))

            if not meta_file.exists():
                self.log_error(f"raw/{folder_name}", "meta.md 파일 없음")
            else:
                self.validate_frontmatter(str(meta_file))
                self.validate_meta_structure(meta_file)

    def validate_topic_structure(self):
        """wiki/topics/ 폴더 구조 검증"""
        topics_dir = self.wiki_root / "wiki" / "topics"
        if not topics_dir.exists():
            self.log_warning("wiki/topics/", "디렉토리가 없음")
            return

        for folder in topics_dir.iterdir():
            if not folder.is_dir():
                continue

            slug = folder.name
            if not self.is_kebab_case(slug):
                self.log_error(f"wiki/topics/{slug}", f"슬러그가 kebab-case 아님")

            # 폴더 노트 구조 검증: wiki/topics/슬러그/슬러그.md
            expected_file = folder / f"{slug}.md"
            if not expected_file.exists():
                self.log_error(f"wiki/topics/{slug}", f"파일 '{slug}.md' 없음 (폴더 노트 구조 필요)")
            else:
                self.validate_frontmatter(str(expected_file))
                self.validate_topic_links(expected_file, slug)

            # 파일이 폴더명과 다르면 경고
            for file in folder.glob("*.md"):
                if file.name != f"{slug}.md" and file.name != "index.md":
                    self.log_warning(f"wiki/topics/{slug}/{file.name}", "폴더명과 일치하지 않는 파일명")

    def validate_frontmatter(self, file_path: str):
        """Frontmatter YAML 유효성 검증"""
        try:
            with open(file_path, 'r', encoding='utf-8') as f:
                content = f.read()
                if content.startswith("---"):
                    # YAML frontmatter 추출
                    end_index = content.find("---", 3)
                    if end_index != -1:
                        frontmatter_text = content[3:end_index].strip()
                        try:
                            yaml.safe_load(frontmatter_text)
                            self.log_info(file_path, "Frontmatter 유효")
                        except yaml.YAMLError as e:
                            self.log_error(file_path, f"Frontmatter YAML 문법 오류: {str(e)[:100]}")
                    else:
                        self.log_error(file_path, "Frontmatter 닫는 --- 없음")
        except Exception as e:
            self.log_error(file_path, f"파일 읽기 오류: {str(e)}")

    def validate_meta_structure(self, meta_file: Path):
        """meta.md 구조 검증"""
        try:
            with open(meta_file, 'r', encoding='utf-8') as f:
                content = f.read()

                # 필수 섹션 확인
                required_sections = ["## 메타정보", "## 한줄 요약"]
                for section in required_sections:
                    if section not in content:
                        self.log_warning(str(meta_file), f"필수 섹션 '{section}' 없음")
        except Exception as e:
            self.log_error(str(meta_file), f"검증 오류: {str(e)}")

    def validate_topic_links(self, topic_file: Path, slug: str):
        """주제 페이지의 링크 검증"""
        try:
            with open(topic_file, 'r', encoding='utf-8') as f:
                content = f.read()

                # 링크 추출: [text](path)
                link_pattern = r'\[([^\]]+)\]\(([^)]+)\)'
                links = re.findall(link_pattern, content)

                for text, link_path in links:
                    # 상대경로 링크만 검증
                    if not link_path.startswith('http'):
                        # 경로 정규화 (Windows 역슬래시 처리)
                        normalized_path = link_path.replace("\\", "/")

                        # 상대경로 해석
                        parts = normalized_path.split("/")
                        current = topic_file.parent
                        try:
                            for part in parts:
                                if part == "..":
                                    current = current.parent
                                elif part and part != ".":
                                    current = current / part

                            full_path = current

                            # 경로 깊이 검증
                            if normalized_path.startswith("../../../raw/"):
                                # raw 참조는 3단계 깊이 OK
                                pass
                            elif normalized_path.startswith("../"):
                                # 형제 토픽 링크
                                if normalized_path.count("../") != 1:
                                    self.log_error(str(topic_file), f"형제 토픽 링크 깊이 오류: {link_path}")
                            elif normalized_path.startswith("../../"):
                                # 다른 경로 (e.g., 결정.md)
                                pass
                            else:
                                self.log_warning(str(topic_file), f"예상치 못한 링크 경로: {link_path}")

                            # 파일 존재 여부 확인
                            if not full_path.exists() and "raw/" not in normalized_path:
                                self.log_error(str(topic_file), f"링크가 가리키는 파일 없음: {link_path}")
                        except Exception as path_error:
                            self.log_warning(str(topic_file), f"경로 계산 오류: {link_path} ({str(path_error)[:50]})")
        except Exception as e:
            self.log_error(str(topic_file), f"링크 검증 오류: {str(e)}")

    def validate_rollup_files(self):
        """결정.md, 액션아이템.md 검증"""
        files_to_check = [
            self.wiki_root / "wiki" / "결정" / "결정.md",
            self.wiki_root / "wiki" / "액션아이템" / "액션아이템.md",
        ]

        for file_path in files_to_check:
            if not file_path.exists():
                self.log_warning(str(file_path), "파일 없음")
                continue

            self.validate_frontmatter(str(file_path))

            # 링크 검증
            try:
                with open(file_path, 'r', encoding='utf-8') as f:
                    content = f.read()
                    link_pattern = r'\[([^\]]+)\]\(([^)]+)\)'
                    links = re.findall(link_pattern, content)

                    for text, link_path in links:
                        if not link_path.startswith('http'):
                            # 경로 정규화
                            normalized_path = link_path.replace("\\", "/")
                            # 현재 파일 위치에서 상대경로 계산
                            full_path = (file_path.parent / normalized_path).resolve()
                            # 심볼릭 링크 해석
                            if not full_path.exists():
                                self.log_error(str(file_path), f"깨진 링크: {link_path}")
            except Exception as e:
                self.log_error(str(file_path), f"검증 오류: {str(e)}")

    def validate_index(self):
        """index.md 검증"""
        index_file = self.wiki_root / "wiki" / "index.md"
        if not index_file.exists():
            self.log_error("wiki/index.md", "파일 없음")
            return

        # 토픽 폴더와 index의 링크 일치 확인
        topics_dir = self.wiki_root / "wiki" / "topics"
        if topics_dir.exists():
            topics = set(f.name for f in topics_dir.iterdir() if f.is_dir())

            try:
                with open(index_file, 'r', encoding='utf-8') as f:
                    content = f.read()

                    for topic_slug in topics:
                        # index에 해당 토픽이 링크되어 있는지 확인
                        if f"topics/{topic_slug}/" not in content:
                            self.log_warning("wiki/index.md", f"토픽 '{topic_slug}'이 index에 없음 (고아 페이지)")
            except Exception as e:
                self.log_error("wiki/index.md", f"검증 오류: {str(e)}")

    def run(self):
        """모든 검증 실행"""
        print("[LINT] HarnessWiki 기계 검증 시작...\n")

        self.validate_raw_structure()
        self.validate_topic_structure()
        self.validate_rollup_files()
        self.validate_index()

        # 결과 출력
        print("\n" + "="*60)
        print("검증 완료")
        print("="*60 + "\n")

        if self.errors:
            print(f"[ERROR] 오류 ({len(self.errors)}건):")
            for error in self.errors:
                print(f"  {error}")
            print()

        if self.warnings:
            print(f"[WARN] 경고 ({len(self.warnings)}건):")
            for warning in self.warnings:
                print(f"  {warning}")
            print()

        if self.info:
            print(f"[INFO] 정보 ({len(self.info)}건):")
            for info_msg in self.info[:5]:  # 처음 5개만 표시
                print(f"  {info_msg}")
            if len(self.info) > 5:
                print(f"  ... 외 {len(self.info) - 5}건")
            print()

        # 요약
        print("[SUMMARY]")
        print(f"  - 오류: {len(self.errors)}건")
        print(f"  - 경고: {len(self.warnings)}건")
        print(f"  - 정보: {len(self.info)}건")

        # 종료 코드
        return 0 if not self.errors else 1


if __name__ == "__main__":
    wiki_root = sys.argv[1] if len(sys.argv) > 1 else "."
    linter = WikiLinter(wiki_root)
    sys.exit(linter.run())

File: scripts/lint.ts
#!/usr/bin/env node
/**
 * HarnessWiki 기계 검증 스크립트 (Lint)
 *
 * 검증 항목:
 * 1. Frontmatter 유효성 (YAML 포맷)
 * 2. 스키마 구조 검증 (경로, 파일명)
 * 3. 파일명 규칙 (kebab-case)
 * 4. 링크 검증 (존재 여부, 경로 깊이)
 */

import fs from 'fs';
import path from 'path';
import { globSync } from 'glob';
import yaml from 'js-yaml';

interface ValidationResult {
  errors: string[];
  warnings: string[];
  info: string[];
}

class WikiLinter {
  private wikiRoot: string;
  private errors: string[] = [];
  private warnings: string[] = [];
  private info: string[] = [];

  constructor(wikiRoot: string = '.') {
    this.wikiRoot = wikiRoot;
  }

  private logError(filePath: string, message: string): void {
    this.errors.push(`[ERROR] ${filePath}: ${message}`);
  }

  private logWarning(filePath: string, message: string): void {
    this.warnings.push(`[WARN]  ${filePath}: ${message}`);
  }

  private logInfo(filePath: string, message: string): void {
    this.info.push(`[INFO]  ${filePath}: ${message}`);
  }

  private isKebabCase(text: string): boolean {
    /**
     * 슬러그가 kebab-case 형식인지 확인
     * 허용: 소문자, 숫자, 하이픈, 한글, 기타 유니코드 문자
     */
    if (!text || text.startsWith('-') || text.endsWith('-')) {
      return false;
    }

    // 한글 포함 확인
    const koreanRegex = /[가-힯]/;
    if (koreanRegex.test(text)) {
      return true; // 한글은 허용
    }

    // 영문+숫자+하이픈만 허용
    return /^[a-z0-9-]+$/.test(text);
  }

  private validateFrontmatter(filePath: string): void {
    try {
      const content = fs.readFileSync(filePath, 'utf-8');

      if (content.startsWith('---')) {
        const endIndex = content.indexOf('---', 3);
        if (endIndex !== -1) {
          const frontmatterText = content.substring(3, endIndex).trim();
          try {
            yaml.load(frontmatterText);
            this.logInfo(filePath, 'Frontmatter 유효');
          } catch (e) {
            const error = e as Error;
            this.logError(
              filePath,
              `Frontmatter YAML 문법 오류: ${error.message.substring(0, 100)}`
            );
          }
        } else {
          this.logError(filePath, 'Frontmatter 닫는 --- 없음');
        }
      }
    } catch (e) {
      const error = e as Error;
      this.logError(filePath, `파일 읽기 오류: ${error.message}`);
    }
  }

  private validateMetaStructure(metaFile: string): void {
    try {
      const content = fs.readFileSync(metaFile, 'utf-8');

      const requiredSections = ['## 메타정보', '## 한줄 요약'];
      for (const section of requiredSections) {
        if (!content.includes(section)) {
          this.logWarning(metaFile, `필수 섹션 '${section}' 없음`);
        }
      }
    } catch (e) {
      const error = e as Error;
      this.logError(metaFile, `검증 오류: ${error.message}`);
    }
  }

  private resolvePath(basePath: string, relativePath: string): string {
    /**
     * 상대경로를 절대경로로 변환
     */
    const parts = relativePath.replace(/\\/g, '/').split('/');
    let current = basePath;

    for (const part of parts) {
      if (part === '..') {
        current = path.dirname(current);
      } else if (part && part !== '.') {
        current = path.join(current, part);
      }
    }

    return current;
  }

  private validateTopicLinks(topicFile: string, slug: string): void {
    try {
      const content = fs.readFileSync(topicFile, 'utf-8');

      // 링크 추출: [text](path)
      const linkPattern = /\[([^\]]+)\]\(([^)]+)\)/g;
      let match;

      while ((match = linkPattern.exec(content)) !== null) {
        const linkPath = match[2];

        // 상대경로 링크만 검증
        if (!linkPath.startsWith('http')) {
          const normalizedPath = linkPath.replace(/\\/g, '/');
          const baseDir = path.dirname(topicFile);
          const fullPath = this.resolvePath(baseDir, normalizedPath);

          // 경로 깊이 검증
          if (normalizedPath.startsWith('../../../raw/')) {
            // raw 참조는 3단계 깊이 OK
          } else if (normalizedPath.startsWith('../')) {
            // 형제 토픽 링크
            const upLevelCount = (normalizedPath.match(/\.\.\//g) || []).length;
            if (upLevelCount !== 1) {
              this.logError(
                topicFile,
                `형제 토픽 링크 깊이 오류: ${linkPath}`
              );
            }
          } else if (normalizedPath.startsWith('../../')) {
            // 다른 경로 (e.g., 결정.md)
          } else {
            this.logWarning(
              topicFile,
              `예상치 못한 링크 경로: ${linkPath}`
            );
          }

          // 파일 존재 여부 확인
          if (!normalizedPath.includes('raw/') && !fs.existsSync(fullPath)) {
            this.logError(topicFile, `링크가 가리키는 파일 없음: ${linkPath}`);
          }
        }
      }
    } catch (e) {
      const error = e as Error;
      this.logError(topicFile, `링크 검증 오류: ${error.message}`);
    }
  }

  private validateRawStructure(): void {
    const rawDir = path.join(this.wikiRoot, 'raw');

    if (!fs.existsSync(rawDir)) {
      this.logWarning('raw/', '디렉토리가 없음');
      return;
    }

    const folders = fs.readdirSync(rawDir);

    for (const folder of folders) {
      const folderPath = path.join(rawDir, folder);

      if (!fs.statSync(folderPath).isDirectory()) {
        continue;
      }

      // YYYY-MM-DD_슬러그 형식 검증
      const match = /^(\d{4})-(\d{2})-(\d{2})_(.+)$/.exec(folder);
      if (!match) {
        this.logError(
          `raw/${folder}`,
          '폴더명 형식 오류: YYYY-MM-DD_슬러그 형식 필요'
        );
        continue;
      }

      const slug = match[4];
      if (!this.isKebabCase(slug)) {
        this.logError(
          `raw/${folder}`,
          `슬러그 '${slug}'가 kebab-case 아님 (허용: 소문자, 숫자, 하이픈, 한글)`
        );
      }

      // raw.md, meta.md 존재 확인
      const rawFile = path.join(folderPath, 'raw.md');
      const metaFile = path.join(folderPath, 'meta.md');

      if (!fs.existsSync(rawFile)) {
        this.logError(`raw/${folder}`, 'raw.md 파일 없음');
      } else {
        this.validateFrontmatter(rawFile);
      }

      if (!fs.existsSync(metaFile)) {
        this.logError(`raw/${folder}`, 'meta.md 파일 없음');
      } else {
        this.validateFrontmatter(metaFile);
        this.validateMetaStructure(metaFile);
      }
    }
  }

  private validateTopicStructure(): void {
    const 주제Dir = path.join(this.wikiRoot, 'wiki', '주제');

    if (!fs.existsSync(주제Dir)) {
      this.logWarning('wiki/주제/', '디렉토리가 없음');
      return;
    }

    const folders = fs.readdirSync(주제Dir);

    for (const folder of folders) {
      const folderPath = path.join(주제Dir, folder);

      if (!fs.statSync(folderPath).isDirectory()) {
        continue;
      }

      const slug = folder;
      if (!this.isKebabCase(slug)) {
        this.logError(`wiki/주제/${slug}`, '슬러그가 kebab-case 아님');
      }

      // 폴더 노트 구조 검증: wiki/주제/슬러그/슬러그.md
      const expectedFile = path.join(folderPath, `${slug}.md`);
      if (!fs.existsSync(expectedFile)) {
        this.logError(
          `wiki/주제/${slug}`,
          `파일 '${slug}.md' 없음 (폴더 노트 구조 필요)`
        );
      } else {
        this.validateFrontmatter(expectedFile);
        this.validateTopicLinks(expectedFile, slug);
      }

      // 파일이 폴더명과 다르면 경고
      const files = fs.readdirSync(folderPath);
      for (const file of files) {
        if (file.endsWith('.md') && file !== `${slug}.md` && file !== 'index.md') {
          this.logWarning(
            `wiki/주제/${slug}/${file}`,
            '폴더명과 일치하지 않는 파일명'
          );
        }
      }
    }
  }

  private validateRollupFiles(): void {
    const filesToCheck = [
      path.join(this.wikiRoot, 'wiki', '결정', '결정.md'),
      path.join(this.wikiRoot, 'wiki', '액션아이템', '액션아이템.md'),
    ];

    for (const filePath of filesToCheck) {
      if (!fs.existsSync(filePath)) {
        this.logWarning(filePath, '파일 없음');
        continue;
      }

      this.validateFrontmatter(filePath);

      // 링크 검증
      try {
        const content = fs.readFileSync(filePath, 'utf-8');
        const linkPattern = /\[([^\]]+)\]\(([^)]+)\)/g;
        let match;

        while ((match = linkPattern.exec(content)) !== null) {
          const linkPath = match[2];

          if (!linkPath.startsWith('http')) {
            const normalizedPath = linkPath.replace(/\\/g, '/');
            const baseDir = path.dirname(filePath);
            const fullPath = this.resolvePath(baseDir, normalizedPath);

            if (!fs.existsSync(fullPath)) {
              this.logError(filePath, `깨진 링크: ${linkPath}`);
            }
          }
        }
      } catch (e) {
        const error = e as Error;
        this.logError(filePath, `검증 오류: ${error.message}`);
      }
    }
  }

  private validateIndex(): void {
    const indexFile = path.join(this.wikiRoot, 'wiki', 'index.md');

    if (!fs.existsSync(indexFile)) {
      this.logError('wiki/index.md', '파일 없음');
      return;
    }

    // 토픽 폴더와 index의 링크 일치 확인
    const 주제Dir = path.join(this.wikiRoot, 'wiki', '주제');
    if (fs.existsSync(주제Dir)) {
      const 주제 = fs
        .readdirSync(주제Dir)
        .filter((f) => fs.statSync(path.join(주제Dir, f)).isDirectory());

      try {
        const content = fs.readFileSync(indexFile, 'utf-8');

        for (const topic of 주제) {
          // index에 해당 토픽이 링크되어 있는지 확인
          if (!content.includes(`주제/${topic}/`)) {
            this.logWarning(
              'wiki/index.md',
              `토픽 '${topic}'이 index에 없음 (고아 페이지)`
            );
          }
        }
      } catch (e) {
        const error = e as Error;
        this.logError('wiki/index.md', `검증 오류: ${error.message}`);
      }
    }
  }

  public run(): number {
    console.log('[LINT] HarnessWiki 기계 검증 시작...\n');

    this.validateRawStructure();
    this.validateTopicStructure();
    this.validateRollupFiles();
    this.validateIndex();

    // 결과 출력
    console.log('\n' + '='.repeat(60));
    console.log('검증 완료');
    console.log('='.repeat(60) + '\n');

    if (this.errors.length > 0) {
      console.log(`[ERROR] 오류 (${this.errors.length}건):`);
      for (const error of this.errors) {
        console.log(`  ${error}`);
      }
      console.log();
    }

    if (this.warnings.length > 0) {
      console.log(`[WARN] 경고 (${this.warnings.length}건):`);
      for (const warning of this.warnings) {
        console.log(`  ${warning}`);
      }
      console.log();
    }

    if (this.info.length > 0) {
      console.log(`[INFO] 정보 (${this.info.length}건):`);
      const displayCount = Math.min(5, this.info.length);
      for (let i = 0; i < displayCount; i++) {
        console.log(`  ${this.info[i]}`);
      }
      if (this.info.length > 5) {
        console.log(`  ... 외 ${this.info.length - 5}건`);
      }
      console.log();
    }

    // 요약
    console.log('[SUMMARY]');
    console.log(`  - 오류: ${this.errors.length}건`);
    console.log(`  - 경고: ${this.warnings.length}건`);
    console.log(`  - 정보: ${this.info.length}건`);

    // 종료 코드
    return this.errors.length > 0 ? 1 : 0;
  }
}

// 실행
const args = process.argv.slice(2);
const wikiRoot = args[0] || '.';
const linter = new WikiLinter(wikiRoot);
process.exit(linter.run());
